### PR TITLE
Implement SasCode::from_str without Regex to avoid costly initial regex build

### PR DESCRIPTION
**Pull request by touilleMan**

### CODE REVIEW

**Comment by vxgmichel** [comment]
I timed the compilation of the regex to be about:
- 13 microseconds in release
- 50 microseconds in WASM
- 200 microseconds in unoptimized
- 400 microseconds in WASM unoptimized

Given the regex is cached using `lazy_static!`, I'm -1 on these kind of error-prone optimizations.

**Comment by vxgmichel** [comment]
Regex are only costly when they use `\w`. I vote for closing this.

File: libparsec/crates/types/src/invite.rs [hunk 1]
```diff
@@ -85,15 +83,15 @@ impl FromStr for SASCode {
...
+            return Err("Invalid SAS code");
         }
-        if PATTERN.is_match(s) {
-            Ok(Self(s.to_string()))
-        } else {
-            Err("Invalid SAS code")
+        for c in s.as_bytes() {
+            if !SAS_CODE_CHARS.contains(c) {
+                return Err("Invalid SAS code");
+            }
```
Comment (FirelightFlagboy) [inline libparsec/crates/types/src/invite.rs:92]: Use iter::any

File: libparsec/crates/types/src/invite.rs [hunk 2]
```diff
@@ -65,7 +64,6 @@ impl ToString for InvitationType {
 
 // SAS code is composed of 4 hexadecimal characters
```
Comment (FirelightFlagboy) [inline libparsec/crates/types/src/invite.rs:65]: The comment seems out of place now, I think it should be move to a doc-comment of the struct `SASCode`

File: libparsec/crates/types/src/invite.rs [hunk 3]
```diff
@@ -65,7 +64,6 @@ impl ToString for InvitationType {
...
 const SAS_CODE_CHARS: &[u8; 32] = b"ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
```
Comment (FirelightFlagboy) [inline libparsec/crates/types/src/invite.rs:66]: Should we add a comment explaining why we don't have `I` (uppercase `i`) and `1` ?

I though we had a comment like that but I don't remember where :thinking: maybe it was lost :shrug:

Comment (touilleMan) [reply libparsec/crates/types/src/invite.rs:66]: I added a comment about that  :+1:

File: libparsec/crates/types/src/invite.rs (hunk 2 above)
Comment (touilleMan) [reply libparsec/crates/types/src/invite.rs:65]: I moved the comment to make it a doc-comment

**Pull request merged**